### PR TITLE
Release appropriate resources when tab is closed

### PR DESCRIPTION
**Pull request by jonathangarelick**
Issue: audio playing in tab continues after tab is closed. Solution: Properly clean up resources that tab holds on close (code taken from Firefox iOS -- was probably removed in an old refactor)

### How was this tested?
- [x] I tested this manually
- [ ] I added automated tests
- [ ] Existing automated tests are enough to cover my changes

### Manual test cases
See below.

### Screenshots and screen recordings
(Remember to unmute video below)

https://user-images.githubusercontent.com/41495718/157348374-0ef93694-6494-4995-9daa-ed6a08fc73b7.mov

## Issues addressed
* https://github.com/neevaco/neeva-ios/issues/3022

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -1363,6 +1363,17 @@ extension BrowserViewController: TabDelegate {
         tab.addContentScript(webuiMessageHelper, name: WebUIMessageHelper.name())
     }
 
+    // Cleans up a tab when it is to be removed.
+    func tab(_ tab: Tab, willDeleteWebView webView: WKWebView) {
+        tab.cancelQueuedAlerts()
+        // Old Firefox iOS code left for reference.
+        //KVOs.forEach { webView.removeObserver(self, forKeyPath: $0.rawValue) }
```
Comment (evancrow) [inline Client/Frontend/Browser/BrowserViewController.swift:1370]: nit: Should have a space after each `//`

File: Client/Frontend/Browser/Tabs/Tab.swift [hunk 2]
```diff
@@ -368,6 +369,14 @@ class Tab: NSObject, ObservableObject {
         //        assertionFailure("web view with URL \(webView.url ?? "(nil)") \(webView) was not deallocated")
         //    }
         //}
```
Comment (evancrow) [inline Client/Frontend/Browser/Tabs/Tab.swift:371]: ditto

**Review by evancrow — APPROVED** [review]
LGTM! (with minor nit)

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -1363,6 +1363,17 @@ extension BrowserViewController: TabDelegate {
...
+        //webView.scrollView.removeObserver(self.scrollController, forKeyPath: KVOConstants.contentSize.rawValue)
+        webView.uiDelegate = nil
+        webView.scrollView.delegate = nil
+        webView.removeFromSuperview()
```
Comment (darinwf) [inline Client/Frontend/Browser/BrowserViewController.swift:1374]: Is it necessary to call `removeFromSuperview()` here? It seems redundant with the call in `Tab.close()`.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -1363,6 +1363,17 @@ extension BrowserViewController: TabDelegate {
         tab.addContentScript(webuiMessageHelper, name: WebUIMessageHelper.name())
     }
 
+    // Cleans up a tab when it is to be removed.
+    func tab(_ tab: Tab, willDeleteWebView webView: WKWebView) {
+        tab.cancelQueuedAlerts()
+        // Old Firefox iOS code left for reference.
```
Comment (darinwf) [inline Client/Frontend/Browser/BrowserViewController.swift:1369]: I think you can just delete these comments. I'm pretty sure we don't need to translate these to the current equivalents.

**Review by darinwf — COMMENTED** [review]
Thanks for the patch! I think we also need to call `Tab.close()` inside `TabManager.removeTabs`. That way the "close all tabs" option will also result in `Tab.close()` being called.

**Comment by jonathangarelick** [comment]
Alright, I believe I've addressed every issue. @darinwf -- I threw in that extra didClose event trigger per your old TODO, since I'm sure you know more about how this works. Let me know if you'd like it taken out.

cc @evancrow

File: Client/Frontend/Browser/Tabs/TabManager/TabManager+CloseTabs.swift [hunk 5]
```diff
@@ -37,11 +37,13 @@ extension TabManager {
...
-        tabsUpdatedPublisher.send()
+        tabsToBeRemoved.forEach { tab in
+            tab.close()
+            TabEvent.post(.didClose, for: tab)
+        }
 
-        // TODO(darin): Don't we need to call Tab.close() on each of the removed tabs, and
-        // don't we need to generate a .didClose TabEvent?
+        tabsUpdatedPublisher.send()
+        storeChanges()
```
Comment (jonathangarelick) [inline Client/Frontend/Browser/Tabs/TabManager/TabManager+CloseTabs.swift:46]: Made some structural changes here to mirror `removeTab`.

Comment (darinwf) [reply Client/Frontend/Browser/Tabs/TabManager/TabManager+CloseTabs.swift:46]: LGTM, thank you!

**Review by darinwf — APPROVED** [review]

**Pull request merged**